pkg/client/rpc_client.go

```
@@ -1360,3 +1532,92 @@ func (r *RPCClient) makeLogValid(log types.Log) (types.Log, error) {
 	log.Index = uint(newIndex)
 	return log, nil
 }
+
+func (r *RPCClient) confidenceToBlockNumber(confidence primitives.ConfidenceLevel) (rpc.BlockNumber, error) {
+	var referencedBlockNumber rpc.BlockNumber
+	switch confidence {
+	case primitives.Finalized:
+		referencedBlockNumber = rpc.FinalizedBlockNumber
+	case primitives.Safe:
+		referencedBlockNumber = rpc.SafeBlockNumber
+	default:
+		return 0, fmt.Errorf("confidence level %s not supported", confidence)
+	}
+
+	if !r.finalityTagEnabled {
+		return rpc.LatestBlockNumber, nil
+	}
+
+	return referencedBlockNumber, nil
+}
+
+func (r *RPCClient) referenceHeadToMaxAvailableHeight(confidence primitives.ConfidenceLevel, referenceHeadHeight int64) (int64, error) {
+	if r.finalityTagEnabled {
```

what if some chains support Finalized but not Safe?

---

Good point. In such cases, I believe we should return an error, as the user should not use the safe tag in this scenario.
If the chain does not support a safe tag (a polygon is the only one that we need to release), it returns an error or, in some cases, a null result. The returned error is self-explanatory. Added specific handling for instances where the returned block is null.